integration-cli/docker_api_swarm_test.go

```diff
@@ -445,6 +445,62 @@ func (s *DockerSwarmSuite) TestApiSwarmServicesStateReporting(c *check.C) {
 		}
 	}
 }
+func (s *DockerSwarmSuite) TestApiSwarmLeaderElection(c *check.C) {
+	// Create 3 nodes
+	d1 := s.AddDaemon(c, true, true)
+	d2 := s.AddDaemon(c, true, true)
+	d3 := s.AddDaemon(c, true, true)
+
+	var leader *SwarmDaemon
+	// save the followers. d1 should be the leader, theoretically, because it
+	// joined first, but that's not guaranteed, i don't think, and we need
+	// to iterate through the followers later until one becomes leader
+	followers := make([]*SwarmDaemon, 0, 3)
+
+	// pick out the leader and the followers
+	for _, d := range []*SwarmDaemon{d1, d2, d3} {
+		if d1.getNode(c, d.NodeID).ManagerStatus.Leader {
+			leader = d
+		} else {
+			followers = append(followers, d)
+		}
+	}
+
+	c.Assert(leader, checker.NotNil)
+	c.Assert(followers, checker.HasLen, 2)
+
+	// stop the leader
+	leader.Stop()
+
+	// wait for an election to occur
+	var newleader *SwarmDaemon
+
+	timeout := time.After(defaultReconciliationTimeout)
+
+	// loop until leader
+	for newleader == nil {
+		select {
+		case <-timeout:
+			break
+		default:
+			for _, d := range followers {
+				if followers[0].getNode(c, d.NodeID).ManagerStatus.Leader {
+					newleader = d
+					break
+				}
+			}
+		}
+	}
+
+	// assert that we have a new leader
+	c.Assert(newleader, checker.NotNil) // if this fails, the loop probably timed out
+
+	// add the old leader back
+	leader.Start()
+
+	// assert that new leader is not the old leader
+	c.Assert(newleader.NodeID, checker.Not(checker.Equals), leader.NodeID)
```

Comment:
I don't think this would ever be possible.
We probably need to wait for the old leader to be available again, then re-grab the status for both the new leader and the old leader... although this is starting to test raft implementation more than docker itself.